### PR TITLE
fix(scale): add opt-in tare lockout during an active shot

### LINKED ISSUE #499: Feature request: tare lockout during shot flow
I was pulling a Filter3 brew this morning, and something sent a tare request in the middle of the shot, which, as the kids would say, "wrecked" my shot. I can't nail down the actual cause, I've enclosed evidence, but it doesn't show much. I'll say that I'm terrible with leaving browser windows open on the skin on multiple machines, but none of those machines were being touched during the shot, including the tablet. Yet, the tare command is there in the log, clear as day, came from somewhere. 

My request is: Could a switch be added in reaprime to disallow a mid-shot tare of the scale? Obviously this would have to be user selectable; in fact, as you may know, there's an active Diaspora thread about the exact opposite, a user wanting to purposely tare during a shot, so clearly, inhibiting taring isn't for everyone. 

This would really be a big help for my flow, I make at least one Filter3 daily and this is the first issue I've run into using it with reaprime. Thanks so much for yopur consideration!

Here's the visualizer link, although I guess it's not needed since the entire shot JSON is in the enclosure. https://visualizer.coffee/shots/5991012a-d7e0-4f93-b107-91d2faf7c8be

[spurious_tare.tgz](https://github.com/user-attachments/files/30374629/spurious_tare.tgz)

### PR DESCRIPTION
**Pull request by bsnelson**
## Summary

> **Naming:** The display name is **Decent.app**. The repo, package, and bundle ID use legacy `reaprime` / `streamline-bridge` — see the naming table in [`CLAUDE.md`](CLAUDE.md).

- Problem: A programmatic `PUT /api/v1/scale/tare` call arriving mid-shot re-zeros the scale and ruins stop-at-weight for the rest of the pour (#499). The reporter's source of the stray tare is unconfirmed, but there's currently no way to prevent it.
- Why it matters: A skin/browser tab, a misbehaving client, or any external caller can silently wreck a shot with no recourse for the user.
- What changed: Adds an opt-in `blockTareDuringShot` setting (default `false`). When enabled, `PUT /api/v1/scale/tare` is rejected with `400` (`type: "block_tare_during_shot"`) while an **app-tracked** espresso shot is actively brewing (`ShotState` other than `idle`/`finished`), and exposes an Enabled/Disabled control for it in the bundled `settings.reaplugin`.
- What did NOT change (scope boundary): Manual tare via the scale's own physical button (never routes through this endpoint), the shot sequencer's own arm-before-pour tare, hot-water stop-at-weight tare, and steam/rinse flows (no `ShotState` concept there) are all unaffected. **`full` gateway mode is an explicit policy exemption**, not an inference that no tracked state can exist there — the skin owns the shot in that mode, so `ScaleHandler` checks `gatewayMode == GatewayMode.full` directly and never blocks a tare there, regardless of any app-tracked (including stale or unexpected) shot state (see Root Cause). No Flutter native-settings-page UI toggle was added — `blockOnNoScale` and `stopHotWaterAtWeight` already set the precedent that these scale-behavior settings are configured via REST/the bundled settings skin, not the native pages.

## Change Type (select all)

- [x] Bug fix
- [ ] Feature
- [ ] Refactor required for the fix
- [ ] Docs
- [ ] Security hardening
- [ ] Chore / infra
- [x] Plugin (DYE2 or bundled skin)

## Scope (select all touched areas)

- [x] REST API / handlers
- [x] Machine state / shot logic
- [x] Scale / weight / flow
- [ ] BLE transport / device comms
- [ ] WebSocket API
- [ ] Profiles / beans / grinders / workflows
- [x] WebUI skins
- [x] Plugins / JS runtime
- [ ] UI / Flutter widgets
- [ ] Storage / Drift database
- [ ] CI / build / infra
- [x] Docs / specs

## Linked Issues

- Closes #499

## Root Cause (if bug fix)

- Root cause: `PUT /api/v1/scale/tare` has no awareness of shot state — any caller can zero the scale mid-pour, and the app has no way to prevent it.
- Missing detection or guardrail: No settings-gated check on shot activity before executing a tare request.
- Contributing context (if known): The reporter (issue author) could not pin down which client issued the stray tare; this PR does not attempt root-cause the specific incident, only add a guardrail some users can opt into (an existing Diaspora thread has users who *want* mid-shot tare, so this is deliberately opt-in, matching the `blockOnNoScale` precedent).
- **Update per review:** the initial version gated only on `ShotState`, reasoning that `De1StateManager` never starts an app-owned `ShotSequencer` in `full` gateway mode. That's true for the common case, but `De1StateManager._handleEspressoState` starts one "regardless of mode" whenever the app's own home screen is foregrounded — so the implicit assumption could have let the lockout misfire under `full` gateway mode in that edge case. `ScaleHandler` now checks `gatewayMode == GatewayMode.full` explicitly alongside the shot-state check, so the exemption holds regardless of that edge case. See `doc/AI_API_NOTES.md` → "Tare Lockout During Shot" for the full writeup.

## Regression Test Plan (if bug fix or refactor)

- Coverage level that should have caught this:
  - [x] Unit test
- Target test or file: `test/services/webserver/scale_handler_tare_lockout_test.dart`
- Scenario the test should lock in: tare is allowed when idle/finished or when the setting is off; tare is rejected with `400`/`block_tare_during_shot` when the setting is on and shot state is preheating/pouring/stopping; tare stays allowed under `full` gateway mode even when the setting is on and a tracked shot looks active.
- If no new test added, why not: N/A — new tests added, plus settings round-trip coverage in `test/webserver/settings_handler_test.dart` and `test/data_export/settings_export_section_test.dart`.

## Documentation Obligations (required)

- [x] API spec updated: `assets/api/rest_v1.yml`
- [x] API docs updated: `doc/Api.md`
- [ ] Plugin docs updated: `doc/Plugins.md`
- [x] Skin docs updated: `doc/Skins.md`
- [ ] Profile docs updated: `doc/Profiles.md`
- [ ] Device docs updated: `doc/DeviceManagement.md`
- [ ] N/A — no docs affected

All four now explicitly state the setting applies only to app-tracked shots and is ignored in `full` gateway mode. Also added `doc/AI_API_NOTES.md` → "Tare Lockout During Shot" recording the design rationale, and updated the archived design doc at `doc/plans/archive/tare-lockout/2026-07-26-issue-499-tare-lockout.md`.

## Security Impact (required)

- New or changed REST endpoints? **Yes** — existing `PUT /api/v1/scale/tare` gains a new conditional `400` response; existing `GET`/`POST /api/v1/settings` gains one new field (`blockTareDuringShot`). No new routes.
- New or changed WebSocket topics? No
- New or changed network calls? No
- BLE/USB surface changed? No
- File system access changed? No
- Plugin sandbox boundary changed? No
- If any `Yes`, explain risk and mitigation: The changed response is opt-in (`blockTareDuringShot` defaults `false`, so existing clients see no behavior change), additive to the response shape (a new possible status code, not a change to the `200` body), and scoped to a single existing endpoint. Risk is limited to a client not handling the new `400`/`block_tare_during_shot` case gracefully — mitigated by documenting it in `doc/Skins.md`/`doc/Api.md` and following the existing `block_no_scale` error-type convention skins already know to handle.

## User-Visible Changes

New settings field `blockTareDuringShot` (boolean, default `false`, via `GET`/`POST /api/v1/settings` and data export/import). When enabled by the user, `PUT /api/v1/scale/tare` returns `400` instead of tareing while an espresso shot is brewing.

### Local gates (run before pushing)

- [x] `flutter analyze` — clean (no new warnings; one pre-existing unrelated warning about a missing plugin asset directory)
- [x] `flutter test` — all 2491 tests pass
- [ ] `(cd packages/dye2-plugin && npm run build)` — not verified locally (local npm toolchain for this package isn't installed in my environment); this PR does not touch `packages/dye2-plugin`, so no impact expected — deferring to CI.

### Manual verification (if applicable)

- OS / platform tested: macOS, `sb-dev.sh start --platform macos --connect-machine MockDe1 --connect-scale MockScale`
- Simulated devices? (`simulate=1`): Yes — full end-to-end smoke test against a running app with `MockDe1` + `MockScale`, in addition to the unit/handler tests using test doubles (`TestScaleController`, `De1Controller.publishShotEvent`).
- Real hardware? (DE1/Bengle/scale): No
- What you personally verified and how: Full `flutter test` suite green; handler tests exercise all five `ShotState` values crossed with the setting on/off plus the full-gateway exemption; settings round-trip (get/set/export/import) verified; live `curl` smoke test against a running simulated instance covering settings round-trip, invalid input, idle tare, blocked mid-shot tare, and the full-gateway exemption (output below).
- Edge cases checked: setting off always allows tare regardless of shot state; setting on still allows tare when idle or finished; blocks during preheating/pouring/stopping; setting on + `full` gateway mode still allows tare mid-shot (the exemption this review requested).
- What you did **not** verify: real DE1/scale hardware.

### Evidence

- [x] Test output (failing before + passing after) — see test run below; `block_tare_during_shot` case did not exist before this change (endpoint had no gate)
- [ ] Log snippets
- [ ] Screenshot / recording (UI changes)
- [x] curl / websocat output (API changes)

```
00:00 +30: ... blockTareDuringShot enabled allows tare once a shot has finished
00:00 +32: ... blockTareDuringShot enabled blocks tare while preheating
00:00 +33: ... blockTareDuringShot enabled blocks tare while pouring
00:00 +34: ... blockTareDuringShot enabled blocks tare while stopping
00:00 +35: ... full gateway mode allows tare even while a tracked shot looks active
00:04 +2491: All tests passed!
```

Live smoke test (`sb-dev.sh` + `curl`, simulated `MockDe1`/`MockScale`):

```
$ curl -sf $BASE/api/v1/settings | jq '{blockTareDuringShot, gatewayMode}'
{ "blockTareDuringShot": false, "gatewayMode": "tracking" }

$ curl -sf -X POST $BASE/api/v1/settings -d '{"blockTareDuringShot": true}'
HTTP 200
$ curl -sf $BASE/api/v1/settings | jq '{blockTareDuringShot}'
{ "blockTareDuringShot": true }

$ curl -X POST $BASE/api/v1/settings -d '{"blockTareDuringShot": "yes"}'
{"message":"blockTareDuringShot must be a boolean"}   HTTP 400

$ curl -X PUT $BASE/api/v1/scale/tare   # idle
HTTP 200

$ curl -X PUT $BASE/api/v1/machine/state/espresso; sleep 4
$ curl -s $BASE/api/v1/machine/state | jq -c '.state'
{"state":"espresso","substate":"preinfusion"}
$ curl -X PUT $BASE/api/v1/scale/tare   # mid-shot, tracking mode
{"details":"Tare blocked: a shot is in progress","type":"block_tare_during_shot"}   HTTP 400
$ curl -X PUT $BASE/api/v1/machine/state/idle   # stop shot

$ curl -X POST $BASE/api/v1/settings -d '{"gatewayMode": "full"}'
$ curl -X PUT $BASE/api/v1/machine/state/espresso; sleep 4
$ curl -s $BASE/api/v1/machine/state | jq -c '.state'
{"state":"espresso","substate":"preinfusion"}
$ curl -X PUT $BASE/api/v1/scale/tare   # mid-shot, full gateway mode
HTTP 200   # exempt, as intended
```

## Compatibility & Migration

- Backward compatible? Yes — new field defaults to `false`/off; existing clients see no behavior change unless they opt in.
- Config / env changes needed? No
- Database migration needed? No
- If any `No` or `Yes`, explain exact steps: N/A

## Risks & Mitigations

- Risk: A skin or client polls/tares aggressively and now silently gets `400`s once a user opts in, without knowing why.
  - Mitigation: Response body carries a stable `type: "block_tare_during_shot"` field (mirrors the existing `block_no_scale` convention) so skins can detect and surface it; documented in `doc/Skins.md` with guidance to show a non-fatal notice rather than retry.

## Update (second review round)

- Corrected the "app never runs a sequencer in full gateway mode" framing everywhere it appeared (`rest_v1.yml`, `doc/Api.md`, `doc/Skins.md`, the bundled settings plugin's help text, and this PR description) to instead describe `full` gateway mode as an explicit policy exemption — the skin owns the shot there, so the lockout never blocks a tare in that mode, regardless of any app-tracked (including stale or unexpected) shot state. `doc/AI_API_NOTES.md` and the archived design doc already had this framing correctly and were left as-is.
- Removed the two inline comments added in `scale_handler.dart` and `scale_handler_tare_lockout_test.dart` — the rationale lives in `doc/AI_API_NOTES.md` per repo convention, so the inline duplicates were redundant.
- Updated Change Type / Scope checkboxes above to reflect the `settings.reaplugin` change (`Plugin (DYE2 or bundled skin)`, `WebUI skins`, `Plugins / JS runtime`).
- Re-ran `flutter analyze` (clean) and `flutter test` (2491 tests, all pass) after these edits.
- Re: CI — the PR-checks run on the latest commit shows `action_required` rather than actually running (visible under this PR's checks / the Actions tab). That looks like it needs a maintainer approval click on the pending workflow run rather than anything I can trigger from a fork PR — let me know if there's something else needed on my end.

### CODE REVIEW

**Review by tadelv — CHANGES_REQUESTED** [review]
Thanks for the contribution — the core settings, persistence, export/import, handler, specification, and regression-test plumbing are generally well structured.

I am requesting changes for two user-facing/API-contract gaps:

1. **Please expose `blockTareDuringShot` in the bundled Settings Viewer.**

The bundled `settings.reaplugin` already exposes comparable controls such as `blockOnNoScale` and `stopHotWaterAtWeight`. Without a control for this new opt-in setting, ordinary users cannot practically enable it without issuing a manual REST request or using a custom skin.

Please add an Enabled/Disabled control in `assets/plugins/settings.reaplugin/plugin.js`, using the existing `updateReaSetting()` path, and update the PR scope accordingly.

2. **Please document the intentional full-gateway exception.**

Full gateway mode deliberately leaves shot ownership to the skin and does not create the app-owned `ShotSequencer`. Therefore `currentShotState` remains outside the tracked active-shot lifecycle, and this tare lockout intentionally does not apply.

That behaviour is acceptable, but the current PR description, OpenAPI specification, and documentation state more generally that tare is rejected whenever an espresso shot is active. Please clarify that the setting is enforced only for app-tracked shots and is ignored in `GatewayMode.full`.

Please update the PR summary/scope boundary, archived design document, OpenAPI description, `doc/Api.md`, `doc/Skins.md`, and preferably `doc/AI_API_NOTES.md`. A regression test should also lock in that tare remains allowed in full gateway mode even when the setting is enabled.

Additional repository-compliance updates:

- The Security Impact section should answer **Yes** for changed REST endpoints because `/api/v1/scale/tare` gains a conditional `400` response and `/api/v1/settings` gains a field.
- Please complete the required `sb-dev.sh` plus `curl` smoke test for the settings round-trip, invalid input, blocked tracked-shot tare, idle tare, and the full-gateway exception.

All GitHub Actions checks are currently passing.

**Comment by bsnelson** [comment]
Thanks for the detailed review — addressed both requested changes plus the compliance items, in `454a2961`:

1. **Settings Viewer control added.** `assets/plugins/settings.reaplugin/plugin.js` now has an Enabled/Disabled select for `blockTareDuringShot`, right after `blockOnNoScale`, using the existing `updateReaSetting()` path.

2. **Full gateway mode exception, made explicit rather than implicit.** You're right that this deserved a closer look: `ScaleHandler` now checks `gatewayMode == GatewayMode.full` directly, rather than relying on the assumption that the app never runs its own `ShotSequencer` in full mode. That assumption mostly holds, but `De1StateManager._handleEspressoState` starts an app-owned `ShotSequencer` "regardless of mode" whenever the app's own home screen is foregrounded — so relying on ShotSequencer-absence alone could have let the lockout misfire under full gateway mode in that edge case. Added a regression test (`scale_handler_tare_lockout_test.dart`) that publishes a `pouring` event under `GatewayMode.full` with the setting on and asserts tare stays allowed. Documented the exemption in the OpenAPI description, `doc/Api.md`, `doc/Skins.md`, the archived design doc, and a new `doc/AI_API_NOTES.md` section ("Tare Lockout During Shot") with the full rationale.

3. **Security Impact section corrected to Yes**, with an explanation of why the risk is scoped/opt-in.

4. **Live smoke test completed** via `sb-dev.sh` + `curl` against a running simulated instance (`MockDe1`/`MockScale`): settings round-trip, invalid input, idle tare, blocked mid-shot tare (`400`/`block_tare_during_shot`, confirmed via `/api/v1/machine/state` showing `preinfusion`), and the full-gateway exemption (same mid-shot state, `gatewayMode: full`, tare returns `200`). Full output is in the updated PR description's Evidence section.

All local gates still green: `flutter analyze` clean, `flutter test` (2491 tests) passing.

**Comment by bsnelson** [comment]
Full disclosure (and I think you know this): The PR and the update were (clearly) authored primarily by Claude, and other than its testing (which I watched in real time, it's fascinating and kudos to you and the team for making the Mock devices and the whole infra to allow testing like this, it's really great!

**Comment by tadelv** [comment]
thanks @bsnelson - indeed most of this PR is our agents talking to one another :D 

Happy to hear the work invested into the infrastructure for testing without real machines is paying off!

What will follow is the formal GPT review

File: assets/api/rest_v1.yml [hunk 1]
```diff
@@ -709,11 +709,36 @@ paths:
...
+      description: |
+        Tares the connected scale. If `blockTareDuringShot` is enabled and an
+        app-tracked espresso shot is actively brewing (not idle/finished), the
+        request is rejected with `400` rather than tared — see
+        `blockTareDuringShot` in `ReaSettings`. In-app tares issued by the shot
+        sequencer itself (e.g. arm-before-pour) are unaffected; only
+        externally-issued tare requests are blocked.
+
+        This only applies to shots the app itself tracks. In `full` gateway
+        mode the skin owns the shot and the app never runs its own shot
```
Comment (tadelv) [inline assets/api/rest_v1.yml]: This explanation contradicts the edge case documented in `doc/AI_API_NOTES.md` and the reason for adding the explicit `GatewayMode.full` check: an app-owned `ShotSequencer` can still be started when the launcher/home screen is foregrounded. Please avoid saying the app “never” runs one or that tracked state cannot exist. Suggested contract wording: full gateway mode is explicitly exempt because the skin owns the shot, regardless of any app-tracked, stale, or unexpected shot state. Please apply the same correction to `doc/Api.md`, `doc/Skins.md`, the Settings Viewer help text, the archived plan, and the PR description.

File: lib/src/services/webserver/scale_handler.dart [hunk 2]
```diff
@@ -2,17 +2,42 @@ part of '../webserver_service.dart';
...
 
   void addRoutes(RouterPlus app) {
     app.put('/api/v1/scale/<command>', (request, command) async {
       switch (command) {
         case 'tare':
           _log.fine("handling api tare command");
+          final shotState = _de1Controller.currentShotState.state;
+          final shotActive =
+              shotState != ShotState.idle && shotState != ShotState.finished;
+          // Full gateway mode leaves shot ownership to the skin — the app
```
Comment (tadelv) [inline lib/src/services/webserver/scale_handler.dart]: Please remove this explanatory comment. `AGENTS.md` says not to add explanatory comments to new code and to place rationale in the matching `doc/AI_*_NOTES.md` file. `isFullGateway` and the condition are clear, and the rationale is already recorded in `doc/AI_API_NOTES.md`.

File: test/services/webserver/scale_handler_tare_lockout_test.dart [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    () {
+      setUp(
+        () => wire(
+          blockTareDuringShot: true,
+          gatewayMode: GatewayMode.full,
+        ),
+      );
+
+      test('allows tare even while a tracked shot looks active', () async {
+        // Full gateway mode leaves shot ownership to the skin; the app does
```
Comment (tadelv) [inline test/services/webserver/scale_handler_tare_lockout_test.dart]: Please remove this explanatory comment for the same `AGENTS.md` rule. The group and test names already describe the required behavior, while the architectural reason belongs in `doc/AI_API_NOTES.md`.

**Review by tadelv — CHANGES_REQUESTED** [review]
Thanks for the update. I re-reviewed the new commit rather than relying on the response summary.

The important functional requests from the first review are now addressed:

- `blockTareDuringShot` is exposed in the bundled Settings Viewer through the existing `updateReaSetting()` path.
- `ScaleHandler` explicitly exempts `GatewayMode.full`.
- A regression test covers an active-looking shot state in full gateway mode.
- The API/security sections and smoke-test evidence were updated.

I am keeping **changes requested** for the following remaining items:

1. **Correct the full-gateway explanation everywhere it appears.** The implementation and `doc/AI_API_NOTES.md` correctly establish that an app-owned `ShotSequencer` can still exist in the launcher/home-screen edge case. Public documentation and UI help text should therefore not say that the app “never” runs a sequencer or that no tracked state can exist. Describe full mode as an explicit policy exemption because the skin owns the shot, regardless of any app-tracked, stale, or unexpected shot state.

2. **Update the PR template scope.** This commit modifies the bundled `settings.reaplugin`, but `Plugin (DYE2 or bundled skin)`, `WebUI skins`, and `Plugins / JS runtime` remain unchecked. At minimum, the plugin change type and plugin/JS scope must be selected; the bundled user-facing settings page also makes `WebUI skins` applicable.

3. **Remove the new explanatory inline comments.** `AGENTS.md` says new code should use clear names and put rationale in the relevant `doc/AI_*_NOTES.md` file. That rationale is already documented in `doc/AI_API_NOTES.md`, so the comments in the handler and regression test are redundant and contrary to the repository rule.

4. **Run CI on the current head before approval.** The current PR-check workflow is `action_required` and created no jobs, so the new commit has not yet received GitHub Actions validation. The submitted local test and smoke-test evidence is useful, but CI should also run and pass.

I did not find a new functional defect in the tare-lockout behavior itself.

**Comment by bsnelson** [comment]
Thanks — addressed all four points:

1. **Wording corrected everywhere.** You're right that "the app never runs a sequencer in full mode" overstated it, given the launcher/home-screen edge case `doc/AI_API_NOTES.md` already documents. Reworded `rest_v1.yml`, `doc/Api.md`, `doc/Skins.md`, and — since you called out UI help text specifically — the bundled `settings.reaplugin`'s help text, to describe `full` gateway mode as an explicit policy exemption (the skin owns the shot, so the lockout never blocks a tare there) rather than an inference that no tracked state can exist. `doc/AI_API_NOTES.md` and the archived design doc already had this right and didn't need changes.
2. **PR template scope corrected** — checked `Plugin (DYE2 or bundled skin)`, `WebUI skins`, and `Plugins / JS runtime`.
3. **Inline comments removed** from `scale_handler.dart` and `scale_handler_tare_lockout_test.dart` — the rationale is in `doc/AI_API_NOTES.md`, so those were redundant per the repo's comment convention.
4. **CI**: the latest commit's PR-check run shows `action_required` rather than actually running — looks like it needs an approval click on your end for the pending workflow run (I don't have permissions to trigger it from a fork PR). Let me know if there's something else needed.

`flutter analyze` clean, `flutter test` (2491 tests) passing after these edits.

**Review by tadelv — APPROVED** [review]
Re-reviewed the latest head (`86d73618`). The second-round findings are fully addressed:

- Full gateway mode is now consistently documented as an explicit policy exemption, without incorrectly claiming tracked shot state cannot exist.
- The two redundant explanatory comments were removed in line with `AGENTS.md`.
- The PR change-type and scope checkboxes now include the bundled plugin/WebUI work.
- The full-gateway regression test remains in place and the handler logic is unchanged and correct.
- PR checks completed successfully: format/analyze/test, Linux build smoke including the DYE2 build, and Python tools tests are all green.

No remaining blocking findings.
